librz/debug/p/native/android_arm.c

```diff
@@ -0,0 +1,659 @@
+			}
+		}
+	}
+
+	/* if we still don't know what to do, we have a problem... */
+	if (reason == RZ_DEBUG_REASON_UNKNOWN) {
+		eprintf("%s: no idea what happened...\n", __func__);
+		reason = RZ_DEBUG_REASON_ERROR;
+	}
+	dbg->reason.tid = pid;
+	dbg->reason.type = reason;
+	return reason;
+}
+#else
+static RzDebugReasonType rz_debug_native_wait(RzDebug *dbg, int pid) {
+	RzDebugReasonType reason = RZ_DEBUG_REASON_UNKNOWN;
+	if (pid == -1) {
+		eprintf("ERROR: rz_debug_native_wait called with pid -1\n");
+		return RZ_DEBUG_REASON_ERROR;
+	}
+
+	reason = linux_dbg_wait(dbg, dbg->tid);
+	dbg->reason.type = reason;
+	return reason;
+}
+#endif
+
+#undef MAXPID
+#define MAXPID 99999
+
+static RzList /*<RzDebugPid *>*/ *rz_debug_native_pids(RzDebug *dbg, int pid) {
+	RzList *list = rz_list_new();
+	if (!list) {
+		return NULL;
+	}
+	return linux_pid_list(pid, list);
+}
+
+RZ_API RZ_OWN RzList /*<RzDebugPid *>*/ *rz_debug_native_threads(RzDebug *dbg, int pid) {
+	RzList *list = rz_list_new();
+	if (!list) {
+		eprintf("No list?\n");
+		return NULL;
+	}
+	return linux_thread_list(dbg, pid, list);
+}
+
+RZ_API ut64 rz_debug_get_tls(RZ_NONNULL RzDebug *dbg, int tid) {
+	rz_return_val_if_fail(dbg, 0);
+	return get_linux_tls_val(dbg, tid);
+}
+
+static int rz_debug_native_reg_read(RzDebug *dbg, int type, ut8 *buf, int size) {
+	if (size < 1) {
+		return false;
+	}
+	return linux_reg_read(dbg, type, buf, size);
+}
+
+static int rz_debug_native_reg_write(RzDebug *dbg, int type, const ut8 *buf, int size) {
+	// XXX use switch or so
+	if (type == RZ_REG_TYPE_DRX) {
+		return false;
+	} else if (type == RZ_REG_TYPE_GPR) {
+		return linux_reg_write(dbg, type, buf, size);
+	} else if (type == RZ_REG_TYPE_FPU) {
+		return linux_reg_write(dbg, type, buf, size);
+	} // else eprintf ("TODO: reg_write_non-gpr (%d)\n", type);
+	return false;
+}
+
+static int io_perms_to_prot(int io_perms) {
+	int prot_perms = PROT_NONE;
+
+	if (io_perms & RZ_PERM_R) {
+		prot_perms |= PROT_READ;
+	}
+	if (io_perms & RZ_PERM_W) {
+		prot_perms |= PROT_WRITE;
+	}
+	if (io_perms & RZ_PERM_X) {
+		prot_perms |= PROT_EXEC;
+	}
+	return prot_perms;
+}
+
+static int linux_map_thp(RzDebug *dbg, ut64 addr, int size) {
+	return false;
+}
+
+static RzDebugMap *linux_map_alloc(RzDebug *dbg, ut64 addr, int size, bool thp) {
+	RzBuffer *buf = NULL;
+	RzDebugMap *map = NULL;
+	char code[1024], *sc_name;
+	int num;
+	/* force to usage of x86.as, not yet working x86.nz */
+	char *asm_list[] = {
+		"x86", "x86.as",
+		"x64", "x86.as",
+		NULL
+	};
+
+	/* NOTE: Since kernel 2.4,  that  system  call  has  been  superseded  by
+		 mmap2(2 and  nowadays  the  glibc  mmap()  wrapper  function invokes
+		 mmap2(2)). If arch is x86_32 then usage mmap2() */
+	if (!strcmp(dbg->arch, "x86") && dbg->bits == 4) {
+		sc_name = "mmap2";
+	} else {
+		sc_name = "mmap";
+	}
+	num = rz_syscall_get_num(dbg->analysis->syscall, sc_name);
+#ifndef MAP_ANONYMOUS
+#define MAP_ANONYMOUS 0x20
+#endif
+	snprintf(code, sizeof(code),
```

Comment:
the task was to split this, so we can refactor for each platform so we can be sure to avoid breaking stuff and remove code that should definitely not be there.